### PR TITLE
2019-12-12-vmx-exit-reasons-and-handlers.md: add post

### PR DESCRIPTION
**Pull request by krystian-hebel**
Signed-off-by: Krystian Hebel <krystian.hebel@3mdeb.com>

### CODE REVIEW

**Comment by miczyg1** [comment]
Looks good to me, however I wonder if it is not too heavy still

**Comment by artur-rs** [comment]
@miczyg1 if looks good to you, then please approve - this post is planned for this week

File: blog/content/post/2020-01-23-vmx-exit-reasons-and-handlers.md [hunk 1]
```diff
@@ -0,0 +1,475 @@
...
+VM exit lands in [exit_handler_entry.asm](https://github.com/Bareflank/hypervisor/blob/ba613e2c687f7042bac6886858cf6da3132a61d6/bfvmm/src/hve/arch/intel_x64/exit_handler_entry.asm)
+initially, where it just saves guest values of registers and calls
+`exit_handler::handle()` from [exit_handler.cpp](https://github.com/Bareflank/hypervisor/blob/ba613e2c687f7042bac6886858cf6da3132a61d6/bfvmm/src/hve/arch/intel_x64/exit_handler.cpp#L727).
+At the point of merging EAPIs to the main tree, this was done with some
+assumptions about name mangling, it is no longer the case in newer code.
+
+Apart from exception handling and dumping CPU state in case of not handled exit
+reason, this function performs two loops:
+
+```
```
Comment (pietrushnic) [inline blog/content/post/2020-01-23-vmx-exit-reasons-and-handlers.md]:
```suggestion
```c
```

File: blog/content/post/2020-01-23-vmx-exit-reasons-and-handlers.md [hunk 2]
```diff
@@ -0,0 +1,475 @@
...
+given port number. If there are no valid handlers for that port (i.e. handlers
+that return `true`), a default one (added with
+`vcpu::add_default_io_instruction_handler()`) is called.
+
+For I/O, user handlers use delegates in form of `bool handler_name(vcpu *, info_t &)`,
+where `info_t` is defined in [io_instruction.h](https://github.com/Bareflank/hypervisor/blob/ba613e2c687f7042bac6886858cf6da3132a61d6/bfvmm/include/hve/arch/intel_x64/vmexit/io_instruction.h#L66)
+as (original comments removed for clarity - check them for description and
+default values; my warnings added instead):
+
+```
```
Comment (pietrushnic) [inline blog/content/post/2020-01-23-vmx-exit-reasons-and-handlers.md]:
```suggestion
```c
```

File: blog/content/post/2020-01-23-vmx-exit-reasons-and-handlers.md [hunk 3]
```diff
@@ -0,0 +1,475 @@
...
+it should use it. As we are building on top of the EFI target from the
+[previous post](https://blog.3mdeb.com/2019/2019-05-15-building-bareflank/),
+adding code to [test_efi.cpp](https://github.com/Bareflank/hypervisor/blob/ba613e2c687f7042bac6886858cf6da3132a61d6/bfvmm/integration/arch/intel_x64/efi/test_efi.cpp)
+is much easier, especially for someone who is not experienced in Cmake.
+
+This file contains a minimal constructor for vCPU. Those two calls are used to
+set up EPT, suffice it to say EPT is required for Bareflank on top of UEFI (more
+about it in the next posts). This is where we are going to add our own handlers:
+
+```
```
Comment (pietrushnic) [inline blog/content/post/2020-01-23-vmx-exit-reasons-and-handlers.md]:
```suggestion
```c
```

File: blog/content/post/2020-01-23-vmx-exit-reasons-and-handlers.md (hunk 1 above)
Comment (pietrushnic) [reply blog/content/post/2020-01-23-vmx-exit-reasons-and-handlers.md]: @artur-rs I'm not sure what is supported by Hugo, but we should have information in README how given code snippets should be marked, good looking code is easier to read.

Comment (pietrushnic) [reply blog/content/post/2020-01-23-vmx-exit-reasons-and-handlers.md]: @artur-rs maybe there is marker for C++?

File: blog/content/post/2020-01-23-vmx-exit-reasons-and-handlers.md [hunk 4]
```diff
@@ -0,0 +1,475 @@
...
+        );
+    }
+```
+
+That wasn't too hard, right? Now, let's move to implementing those delegates,
+starting with the read handler. We need to do so even though we won't do
+anything in this handler, as there is no way to enable exiting on writes, but
+not on reads for the given port. Delegates can be private members of vcpu class:
+
+```
```
Comment (pietrushnic) [inline blog/content/post/2020-01-23-vmx-exit-reasons-and-handlers.md]:
```suggestion
```c
```

File: blog/content/post/2020-01-23-vmx-exit-reasons-and-handlers.md [hunk 5]
```diff
@@ -0,0 +1,475 @@
...
+That's it, thanks to sane default values of `info_t` - neither read nor advance
+is ignored, so the instruction behaves as expected.
+
+### First modification - no output
+
+This one is also easy. Remember `write_value` in `info_t`? Just set it to `true`
+and we're done. The value will not be sent through this port anymore from the VM
+(Bareflank can still print its messages with e.g. `bfdebug_info()`).
+
+```
```
Comment (pietrushnic) [inline blog/content/post/2020-01-23-vmx-exit-reasons-and-handlers.md]:
```suggestion
```c
```

File: blog/content/post/2020-01-23-vmx-exit-reasons-and-handlers.md [hunk 6]
```diff
@@ -0,0 +1,475 @@
...
+
+To test it, follow instructions from the [previous post](https://blog.3mdeb.com/2019/2019-05-15-building-bareflank/).
+Compare output from VGA and serial.
+
+### Second modification - case swap
+
+This one seems easier than it is, actually. Starting with a naive approach:
+
+```
```
Comment (pietrushnic) [inline blog/content/post/2020-01-23-vmx-exit-reasons-and-handlers.md]:
```suggestion
```c
```

File: blog/content/post/2020-01-23-vmx-exit-reasons-and-handlers.md [hunk 7]
```diff
@@ -0,0 +1,475 @@
...
+The issue is that UEFI uses [ANSI escape codes](https://en.wikipedia.org/wiki/ANSI_escape_code)
+to move around the screen. From table with [terminal output sequences](https://en.wikipedia.org/wiki/ANSI_escape_code#Terminal_output_sequences)
+we can read that those sequences start with `ESC [`, followed by some non-alpha
+sequence (zero or more decimal numbers and possibly a semicolon), followed by a
+single letter. This means that we must pass everything from the initial `0x1B`
+byte (`ESC`) up to and including first letter character unchanged.
+
+At the very beginning of `out_handler()` add:
+
+```
```
Comment (pietrushnic) [inline blog/content/post/2020-01-23-vmx-exit-reasons-and-handlers.md]:
```suggestion
```c
```

File: blog/content/post/2020-01-23-vmx-exit-reasons-and-handlers.md (hunk 1 above)
Comment (artur-rs) [reply blog/content/post/2020-01-23-vmx-exit-reasons-and-handlers.md]: @pietrushnic https://gohugo.io/content-management/syntax-highlighting/#list-of-chroma-highlighting-languages

Markers for C++: 
```
C, CPP, H, c++, cc, cp, cpp, cxx, h++, hh, hpp, hxx
```

I will add this to README as well

Comment (krystian-hebel) [reply blog/content/post/2020-01-23-vmx-exit-reasons-and-handlers.md]: All markers added

**Pull request merged**